File: lib/graphql/sdk-python/nanobanana_sdk/logger.py
"""
艹！Nano Banana GraphQL SDK 日志记录模块

这个SB模块提供结构化日志记录，方便你tm调试问题！
"""

import logging
import json
from typing import Any, Dict, Optional
from datetime import datetime


class SDKLogger:
    """
    艹！SDK 日志记录器

    提供结构化日志记录，包含：
    - 时间戳
    - 操作名称
    - 变量
    - 错误信息
    - 响应时间等
    """

    def __init__(
        self,
        name: str = "nanobanana_sdk",
        level: int = logging.INFO,
        enable_logging: bool = True,
    ):
        """
        初始化日志记录器

        Args:
            name: 日志记录器名称
            level: 日志级别（DEBUG, INFO, WARNING, ERROR, CRITICAL）
            enable_logging: 是否启用日志（默认 True）
        """
        self.logger = logging.getLogger(name)
        self.logger.setLevel(level)
        self.enable_logging = enable_logging

        # 如果没有 handler，添加一个控制台 handler
        if not self.logger.handlers:
            handler = logging.StreamHandler()
            handler.setLevel(level)

            # 老王风格的日志格式
            formatter = logging.Formatter(
                "[%(asctime)s] [%(name)s] [%(levelname)s] %(message)s",
                datefmt="%Y-%m-%d %H:%M:%S"
            )
            handler.setFormatter(formatter)
            self.logger.addHandler(handler)

    def _format_log_data(self, data: Dict[str, Any]) -> str:
        """
        格式化日志数据

        Args:
            data: 要记录的数据字典

        Returns:
            格式化后的字符串
        """
        try:
            return json.dumps(data, ensure_ascii=False, indent=2)
        except Exception:
            return str(data)

    def debug(self, message: str, **kwargs):
        """DEBUG 级别日志"""
        if self.enable_logging:
            self.logger.debug(message, extra=kwargs)

    def info(self, message: str, **kwargs):
        """INFO 级别日志"""
        if self.enable_logging:
            self.logger.info(message, extra=kwargs)

    def warning(self, message: str, **kwargs):
        """WARNING 级别日志"""
        if self.enable_logging:
            self.logger.warning(message, extra=kwargs)

    def error(self, message: str, **kwargs):
        """ERROR 级别日志"""
        if self.enable_logging:
            self.logger.error(message, extra=kwargs)

    def critical(self, message: str, **kwargs):
        """CRITICAL 级别日志"""
        if self.enable_logging:
            self.logger.critical(message, extra=kwargs)

    def log_request(
        self,
        operation_name: str,
        variables: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
    ):
        """
        记录 GraphQL 请求

        Args:
            operation_name: 操作名称
            variables: 变量
            headers: 请求头
        """
        if not self.enable_logging:
            return

        log_data = {
            "type": "REQUEST",
            "operation": operation_name,
            "timestamp": datetime.now().isoformat(),
        }

        if variables:
            log_data["variables"] = variables

        if headers:
            # 隐藏敏感信息
            safe_headers = {
                k: v if k.lower() not in ["authorization", "cookie"] else "***"
                for k, v in headers.items()
            }
            log_data["headers"] = safe_headers

        self.info(f"发起请求: {operation_name}")
        self.debug(self._format_log_data(log_data))

    def log_response(
        self,
        operation_name: str,
        duration_ms: float,
        success: bool = True,
        error: Optional[Exception] = None,
    ):
        """
        记录 GraphQL 响应

        Args:
            operation_name: 操作名称
            duration_ms: 响应时间（毫秒）
            success: 是否成功
            error: 错误对象（如果失败）
        """
        if not self.enable_logging:
            return

        log_data = {
            "type": "RESPONSE",
            "operation": operation_name,
            "duration_ms": round(duration_ms, 2),
            "success": success,
            "timestamp": datetime.now().isoformat(),
        }

        if error:
            log_data["error"] = {
                "type": type(error).__name__,
                "message": str(error),
            }

        if success:
            self.info(f"请求成功: {operation_name} ({duration_ms:.2f}ms)")
        else:
            self.error(f"请求失败: {operation_name} ({duration_ms:.2f}ms)")

        self.debug(self._format_log_data(log_data))

    def log_retry(
        self,
        operation_name: str,
        attempt: int,
        max_attempts: int,
        delay: float,
        error: Exception,
    ):
        """
        记录重试信息

        Args:
            operation_name: 操作名称
            attempt: 当前尝试次数
            max_attempts: 最大尝试次数
            delay: 延迟时间（秒）
            error: 导致重试的错误
        """
        if not self.enable_logging:
            return

        log_data = {
            "type": "RETRY",
            "operation": operation_name,
            "attempt": attempt,
            "max_attempts": max_attempts,
            "delay_seconds": delay,
            "error": {
                "type": type(error).__name__,
                "message": str(error),
            },
            "timestamp": datetime.now().isoformat(),
        }

        self.warning(
            f"重试中 ({attempt}/{max_attempts}): {operation_name} "
            f"(延迟 {delay:.2f}秒)"
        )
        self.debug(self._format_log_data(log_data))


# 默认日志记录器实例
default_logger = SDKLogger()


# 便捷函数
def set_log_level(level: int):
    """
    设置日志级别

    Args:
        level: logging.DEBUG, logging.INFO, logging.WARNING, logging.ERROR, logging.CRITICAL
    """
    default_logger.logger.setLevel(level)


def enable_logging(enabled: bool = True):
    """
    启用或禁用日志

    Args:
        enabled: True 表示启用，False 表示禁用
    """
    default_logger.enable_logging = enabled


def get_logger(name: str = "nanobanana_sdk") -> SDKLogger:
    """
    获取日志记录器实例

    Args:
        name: 日志记录器名称

    Returns:
        SDKLogger 实例
    """
    return SDKLogger(name=name)

File: lib/graphql/sdk-go/logger.go
// 艹！Nano Banana GraphQL SDK for Go - 日志模块
//
// 这个SB文件实现了结构化日志系统！
//
// 主要功能：
// - 结构化日志输出（JSON格式）
// - 多日志级别（DEBUG/INFO/WARN/ERROR）
// - 敏感数据脱敏（Authorization、Cookie等header）
// - 请求/响应日志
// - 重试日志
// - 性能指标记录

package nanobanana

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"
	"time"
)

// LogLevel 日志级别
type LogLevel int

const (
	// LogLevelDebug DEBUG级别（最详细）
	LogLevelDebug LogLevel = iota
	// LogLevelInfo INFO级别（一般信息）
	LogLevelInfo
	// LogLevelWarn WARN级别（警告信息）
	LogLevelWarn
	// LogLevelError ERROR级别（错误信息）
	LogLevelError
)

// String 返回日志级别的字符串表示
func (l LogLevel) String() string {
	switch l {
	case LogLevelDebug:
		return "DEBUG"
	case LogLevelInfo:
		return "INFO"
	case LogLevelWarn:
		return "WARN"
	case LogLevelError:
		return "ERROR"
	default:
		return "UNKNOWN"
	}
}

// ParseLogLevel 解析日志级别字符串
func ParseLogLevel(level string) LogLevel {
	switch strings.ToUpper(level) {
	case "DEBUG":
		return LogLevelDebug
	case "INFO":
		return LogLevelInfo
	case "WARN", "WARNING":
		return LogLevelWarn
	case "ERROR":
		return LogLevelError
	default:
		return LogLevelInfo
	}
}

// SDKLogger SDK日志记录器
type SDKLogger struct {
	// name 日志记录器名称
	name string

	// level 当前日志级别
	level LogLevel

	// enabled 是否启用日志
	enabled bool

	// logger 底层logger
	logger *log.Logger
}

// NewSDKLogger 创建新的日志记录器
//
// 参数：
//   - name: 日志记录器名称（默认："nanobanana_sdk"）
//   - level: 日志级别（默认：INFO）
//   - enabled: 是否启用日志（默认：true）
//
// 返回：
//   - *SDKLogger: 日志记录器实例
func NewSDKLogger(name string, level LogLevel, enabled bool) *SDKLogger {
	if name == "" {
		name = "nanobanana_sdk"
	}

	return &SDKLogger{
		name:    name,
		level:   level,
		enabled: enabled,
		logger:  log.New(os.Stdout, "", 0), // 不使用默认前缀和标志
	}
}

// SetLevel 设置日志级别
func (l *SDKLogger) SetLevel(level LogLevel) {
	l.level = level
}

// SetEnabled 设置是否启用日志
func (l *SDKLogger) SetEnabled(enabled bool) {
	l.enabled = enabled
}

// IsEnabled 检查日志是否启用
func (l *SDKLogger) IsEnabled() bool {
	return l.enabled
}

// shouldLog 检查是否应该记录该级别的日志
func (l *SDKLogger) shouldLog(level LogLevel) bool {
	return l.enabled && level >= l.level
}

// log 记录日志（内部方法）
func (l *SDKLogger) log(level LogLevel, message string, fields map[string]interface{}) {
	if !l.shouldLog(level) {
		return
	}

	// 构建日志对象
	logData := map[string]interface{}{
		"timestamp": time.Now().Format(time.RFC3339),
		"level":     level.String(),
		"logger":    l.name,
		"message":   message,
	}

	// 合并额外字段
	for k, v := range fields {
		logData[k] = v
	}

	// JSON序列化
	jsonData, err := json.Marshal(logData)
	if err != nil {
		// 如果序列化失败，输出原始消息
		l.logger.Printf("[%s] %s (JSON marshal error: %v)", level.String(), message, err)
		return
	}

	// 输出JSON日志
	l.logger.Println(string(jsonData))
}

// Debug 记录DEBUG级别日志
func (l *SDKLogger) Debug(message string, fields map[string]interface{}) {
	l.log(LogLevelDebug, message, fields)
}

// Info 记录INFO级别日志
func (l *SDKLogger) Info(message string, fields map[string]interface{}) {
	l.log(LogLevelInfo, message, fields)
}

// Warn 记录WARN级别日志
func (l *SDKLogger) Warn(message string, fields map[string]interface{}) {
	l.log(LogLevelWarn, message, fields)
}

// Error 记录ERROR级别日志
func (l *SDKLogger) Error(message string, fields map[string]interface{}) {
	l.log(LogLevelError, message, fields)
}

// RedactHeaders 脱敏HTTP headers（隐藏敏感信息）
//
// 敏感header列表：
// - Authorization（认证token）
// - Cookie（会话cookie）
// - X-API-Key（API密钥）
// - X-Auth-Token（认证token）
//
// 参数：
//   - headers: 原始headers
//
// 返回：
//   - map[string]string: 脱敏后的headers
func RedactHeaders(headers map[string]string) map[string]string {
	if headers == nil {
		return nil
	}

	sensitiveKeys := []string{
		"authorization",
		"cookie",
		"x-api-key",
		"x-auth-token",
	}

	redacted := make(map[string]string)
	for k, v := range headers {
		keyLower := strings.ToLower(k)
		isSensitive := false

		for _, sensitive := range sensitiveKeys {
			if keyLower == sensitive {
				isSensitive = true
				break
			}
		}

		if isSensitive {
			redacted[k] = "***"
		} else {
			redacted[k] = v
		}
	}

	return redacted
}

// LogRequest 记录GraphQL请求日志
//
// 参数：
//   - operationName: 操作名称
//   - variables: 请求变量
//   - headers: 请求headers
func (l *SDKLogger) LogRequest(
	operationName string,
	variables map[string]interface{},
	headers map[string]string,
) {
	if !l.shouldLog(LogLevelInfo) {
		return
	}

	// 脱敏headers
	safeHeaders := RedactHeaders(headers)

	fields := map[string]interface{}{
		"type":           "REQUEST",
		"operation_name": operationName,
		"variables":      variables,
		"headers":        safeHeaders,
	}

	l.Info(fmt.Sprintf("发起GraphQL请求: %s", operationName), fields)
}

// LogResponse 记录GraphQL响应日志
//
// 参数：
//   - operationName: 操作名称
//   - durationMs: 请求耗时（毫秒）
//   - statusCode: HTTP状态码（可选，0表示无）
//   - hasError: 是否有错误
func (l *SDKLogger) LogResponse(
	operationName string,
	durationMs int64,
	statusCode int,
	hasError bool,
) {
	if !l.shouldLog(LogLevelInfo) {
		return
	}

	fields := map[string]interface{}{
		"type":           "RESPONSE",
		"operation_name": operationName,
		"duration_ms":    durationMs,
		"has_error":      hasError,
	}

	if statusCode > 0 {
		fields["status_code"] = statusCode
	}

	message := fmt.Sprintf("GraphQL请求完成: %s (耗时: %dms)", operationName, durationMs)
	if hasError {
		l.Warn(message, fields)
	} else {
		l.Info(message, fields)
	}
}

// LogRetry 记录重试日志
//
// 参数：
//   - operationName: 操作名称
//   - attempt: 当前尝试次数
//   - maxAttempts: 最大尝试次数
//   - err: 错误对象
//   - delayMs: 延迟时间（毫秒）
func (l *SDKLogger) LogRetry(
	operationName string,
	attempt int,
	maxAttempts int,
	err *GraphQLSDKError,
	delayMs int64,
) {
	if !l.shouldLog(LogLevelWarn) {
		return
	}

	fields := map[string]interface{}{
		"type":           "RETRY",
		"operation_name": operationName,
		"attempt":        attempt,
		"max_attempts":   maxAttempts,
		"error_type":     err.ErrorType,
		"error_message":  err.Message,
		"retry_delay_ms": delayMs,
	}

	l.Warn(fmt.Sprintf("请求失败，准备重试 (尝试 %d/%d)", attempt, maxAttempts), fields)
}

// LogError 记录错误日志（带完整错误信息）
//
// 参数：
//   - operationName: 操作名称
//   - err: 错误对象
func (l *SDKLogger) LogError(operationName string, err *GraphQLSDKError) {
	if !l.shouldLog(LogLevelError) {
		return
	}

	fields := map[string]interface{}{
		"type":           "ERROR",
		"operation_name": operationName,
		"error_type":     err.ErrorType,
		"error_message":  err.Message,
		"retryable":      err.Retryable,
		"timestamp":      err.Timestamp.Format(time.RFC3339),
	}

	if err.StatusCode > 0 {
		fields["status_code"] = err.StatusCode
	}

	if err.Variables != nil {
		fields["variables"] = err.Variables
	}

	if err.Extensions != nil {
		fields["extensions"] = err.Extensions
	}

	l.Error(fmt.Sprintf("GraphQL请求失败: %s", operationName), fields)
}

// LogPerformance 记录性能指标日志
//
// 参数：
//   - operationName: 操作名称
//   - metrics: 性能指标
func (l *SDKLogger) LogPerformance(operationName string, metrics map[string]interface{}) {
	if !l.shouldLog(LogLevelDebug) {
		return
	}

	fields := map[string]interface{}{
		"type":           "PERFORMANCE",
		"operation_name": operationName,
	}

	// 合并性能指标
	for k, v := range metrics {
		fields[k] = v
	}

	l.Debug(fmt.Sprintf("性能指标: %s", operationName), fields)
}
